subsys/bluetooth/audio/cap_commander.c

```diff
@@ -724,15 +734,236 @@ int bt_cap_commander_change_volume_offset(
+		const struct bt_cap_common_client *client =
+			bt_cap_common_get_client(param->type, member);
+		struct bt_micp_mic_ctlr *mic_ctlr;
+		struct bt_micp_included included;
+		int err;
+
+		if (client == NULL) {
+			LOG_DBG("Invalid param->param[%zu].member", i);
+			return false;
+		}
+
+		mic_ctlr = bt_micp_mic_ctlr_get_by_conn(client->conn);
+		if (mic_ctlr == NULL) {
+			LOG_DBG("Microphone control not available for param->param[%zu].member", i);
+			return false;
+		}
+
+		err = bt_micp_mic_ctlr_included_get(mic_ctlr, &included);
+		if (err != 0 || included.aics_cnt == 0) {
+			LOG_DBG("Microphone audio input control not available for "
+				"param->param[%zu].member",
+				i);
+			return -ENOEXEC;
+		}
+
+		for (size_t j = 0U; j < i; j++) {
+			const union bt_cap_set_member *other = &param->param[j].member;
+
+			if (other == member) {
+				LOG_DBG("param->param[%zu].member (%p) is duplicated by "
+					"param->param[%zu].member (%p)",
+					j, other, i, member);
+				return false;
+			}
+		}
+	}
+
+	return true;
+}
+
+static void cap_commander_micp_gain_set_cb(struct bt_aics *inst, int err)
+{
+	struct bt_cap_common_proc *active_proc = bt_cap_common_get_active_proc();
+	struct bt_conn *conn;
+	int micp_err;
+
+	LOG_DBG("bt_aics %p", (void *)inst);
+
+	micp_err = bt_aics_client_conn_get(inst, &conn);
+	if (micp_err != 0) {
+		LOG_ERR("Failed to get conn by aics: %d", micp_err);
+		return;
+	}
+
+	LOG_DBG("conn %p", (void *)conn);
+	if (!bt_cap_common_conn_in_active_proc(conn)) {
+		/* State change happened outside of a procedure; ignore */
+		return;
+	}
+
+	if (err != 0) {
+		LOG_DBG("Failed to set gain: %d", err);
+		bt_cap_common_abort_proc(conn, err);
+	} else {
+		active_proc->proc_done_cnt++;
+
+		LOG_DBG("Conn %p gain updated (%zu/%zu streams done)", (void *)conn,
+			active_proc->proc_done_cnt, active_proc->proc_cnt);
+	}
+
+	if (bt_cap_common_proc_is_aborted()) {
+		LOG_DBG("Proc is aborted");
+		if (bt_cap_common_proc_all_handled()) {
+			LOG_DBG("All handled");
+			cap_commander_unicast_audio_proc_complete();
+		}
+
+		return;
+	}
+
+	if (!bt_cap_common_proc_is_done()) {
+		const struct bt_cap_commander_proc_param *proc_param;
+
+		proc_param = &active_proc->proc_param.commander[active_proc->proc_done_cnt];
+		conn = proc_param->conn;
+		active_proc->proc_initiated_cnt++;
+		err = bt_aics_gain_set(proc_param->change_gain.aics, proc_param->change_gain.gain);
+		if (err != 0) {
+			LOG_DBG("Failed to set gain for conn %p: %d", (void *)conn, err);
+			bt_cap_common_abort_proc(conn, err);
+			cap_commander_unicast_audio_proc_complete();
+		}
+	} else {
+		cap_commander_unicast_audio_proc_complete();
+	}
+}
+
 int bt_cap_commander_change_microphone_gain_setting(
 	const struct bt_cap_commander_change_microphone_gain_setting_param *param)
 {
+	const struct bt_cap_commander_proc_param *proc_param;
+	struct bt_cap_common_proc *active_proc;
+	struct bt_conn *conn;
+	int err;
 
-	return -ENOSYS;
+	if (bt_cap_common_proc_is_active()) {
```

Comment:
Since `bt_cap_common_proc_is_active` only checks if proc is active, and `bt_cap_common_start_proc` does not check this at all, there is always a risk that another process hijacks the process in between these calls.

We previously talked about a lot of the Audio API needs mutex guarding against this sort of things, so I guess we can leave this for future? 

I just wanted to point this out

Reply:
Indeed. There is a potential race condition here :)